File: backend/bims-BE/app/schemas/campaign.py
"""
Campaign Schemas aligned with UML.
"""
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional, List
from app.models.campaign import CampaignStatus, CampaignContentType


class PlatformBrief(BaseModel):
    """Platform id and name for campaign card/details."""
    id: str
    name: str


class CampaignBase(BaseModel):
    title: str
    category: Optional[str] = None
    # 0 = face creator campaign, 1 = faceless creator campaign
    campaign_type: int = Field(..., ge=0, le=1)
    content_type: CampaignContentType
    description: Optional[str] = None
    total_budget: float
    rate_per_million_views: float
    max_submissions_per_account: Optional[int] = None
    max_earnings_per_creator: Optional[float] = None
    max_earnings_per_post: Optional[float] = None
    start_date: date
    end_date: date
    logo_drive_link: Optional[str] = None
    guidelines_link: Optional[str] = None
    discord_link: Optional[str] = None


class CampaignCreate(CampaignBase):
    """Create campaign. Include platform_ids to link campaign to platforms. Admin may set brand_id to create on behalf of a brand."""
    platform_ids: List[str] = []
    brand_id: Optional[str] = None  # Only used when caller is admin; otherwise ignored


class CampaignUpdate(BaseModel):
    title: Optional[str] = None
    category: Optional[str] = None
    campaign_type: Optional[int] = Field(None, ge=0, le=1)
    content_type: Optional[CampaignContentType] = None
    description: Optional[str] = None
    total_budget: Optional[float] = None
    used_budget: Optional[float] = None
    rate_per_million_views: Optional[float] = None
    max_submissions_per_account: Optional[int] = None
    max_earnings_per_creator: Optional[float] = None
    max_earnings_per_post: Optional[float] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: Optional[CampaignStatus] = None
    logo_drive_link: Optional[str] = None
    guidelines_link: Optional[str] = None
    discord_link: Optional[str] = None
    platform_ids: Optional[List[str]] = None  # If set, replaces campaign's linked platforms


class CampaignResponse(CampaignBase):
    id: str
    brand_id: str
    status: CampaignStatus
    used_budget: float
    created_at: datetime
    platforms: List[PlatformBrief] = []  # Platforms linked to this campaign (for card/details)
    participant_count: int = 0  # Total creators joined (participated) in this campaign
    participant_avatars: List[str] = []  # Up to 5 profile picture URLs for campaign card avatar stack

    class Config:
        from_attributes = True

File: frontend/clipify/types/campaign.ts
export type CampaignStatus = "ACTIVE" | "PAUSED" | "COMPLETED";

export type CampaignContentType = "VIDEO" | "IMAGE";

/** 0 = face creators (apply → approve); 1 = faceless (submit-link flow) */
export type CampaignType = 0 | 1;

export interface Platform {
  id: string;
  name: string;
}

export interface Campaign {
  id: string;
  brand_id: string;
  title: string;
  /** 0 face (apply flow), 1 faceless (submit link). Older APIs may omit — treat as 0. */
  campaign_type?: CampaignType;
  category?: string;
  content_type: CampaignContentType;
  description?: string;
  total_budget: number;
  used_budget: number;
  rate_per_million_views: number;
  max_submissions_per_account?: number;
  max_earnings_per_creator?: number;
  max_earnings_per_post?: number;
  start_date: string;
  end_date: string;
  status: CampaignStatus;
  logo_drive_link?: string;
  guidelines_link?: string;
  discord_link?: string;
  created_at: string;
  platforms?: Platform[];
  /** Number of participants (creators) in the campaign */
  participant_count?: number;
  /** Avatar URLs for participants (e.g. for display on cards/details) */
  participant_avatars?: string[];
}

export type CampaignCreatePayload = Omit<
  Campaign,
  "id" | "brand_id" | "used_budget" | "status" | "created_at" | "platforms" | "campaign_type"
> & {
  platform_ids: string[];
  brand_id?: string; // only when admin creates on behalf of a brand
  campaign_type: CampaignType;
};

export type CampaignUpdatePayload = Partial<CampaignCreatePayload> & {
  status?: CampaignStatus; // admin/brand can update status via PUT
};

/** Normalize API responses that omit `campaign_type` (legacy). */
export function getCampaignType(campaign: Pick<Campaign, "campaign_type">): CampaignType {
  const t = campaign.campaign_type;
  return t === 1 ? 1 : 0;
}

